kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoKVTable.java

```diff
@@ -0,0 +1,156 @@
+import org.bson.codecs.configuration.CodecProvider;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.codecs.pojo.PojoCodecProvider;
+import org.bson.types.ObjectId;
+
+public class MongoKVTable implements RemoteKVTable {
+
+  private final String name;
+  private final MongoCollection<KVDoc> collection;
+  private final MongoCollection<MetadataDoc> metadata;
+
+  private final ConcurrentMap<Integer, ObjectId> metadataRows = new ConcurrentHashMap<>();
+
+  public MongoKVTable(final MongoClient client, final String name) {
+    this.name = name;
+    CodecProvider pojoCodecProvider = PojoCodecProvider.builder().automatic(true).build();
+    CodecRegistry pojoCodecRegistry = fromRegistries(getDefaultCodecRegistry(), fromProviders(pojoCodecProvider));
+
+    final MongoDatabase database = client.getDatabase(name).withCodecRegistry(pojoCodecRegistry);
+    collection = database.getCollection(name, KVDoc.class);
+    metadata = database.getCollection(name, MetadataDoc.class);
+  }
+
+  @Override
+  public String name() {
+    return name;
+  }
+
+  @Override
+  public WriterFactory<Bytes> init(final SubPartitioner partitioner, final int kafkaPartition) {
+    metadataRows.computeIfAbsent(kafkaPartition, kp -> {
+      final UpdateResult updateResult = metadata.updateOne(
+          Filters.eq("partition", kafkaPartition),
+          Updates.setOnInsert("offset", NO_COMMITTED_OFFSET),
+          new UpdateOptions().upsert(true));
+
+      if (updateResult.getUpsertedId() != null) {
+        return updateResult.getUpsertedId().asObjectId().getValue();
+      }
+
+      // find existing one
+      final MetadataDoc metadataPartition = metadata.find(
+          Filters.eq("partition", kafkaPartition)
+      ).first();
+      if (metadataPartition == null) {
+        throw new IllegalStateException("No metadata partition despite upsert");
+      }
+      return metadataPartition.id();
+    });
+
+    return new MongoWriterFactory<>(this);
+  }
+
+  @Override
+  public byte[] get(final int partition, final Bytes key, final long minValidTs) {
+    final KVDoc v = collection.find(Filters.eq("key", key.get())).first();
+    return v == null ? null : v.getValue();
+  }
+
+  @Override
+  public KeyValueIterator<Bytes, byte[]> range(final int partition, final Bytes from,
+                                               final Bytes to,
+                                               final long minValidTs) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public KeyValueIterator<Bytes, byte[]> all(final int partition, final long minValidTs) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public BoundStatement insert(
+      final int partitionKey,
+      final Bytes key,
+      final byte[] value,
+      final long epochMillis
+  ) {
+    collection.updateOne(
+        Filters.eq("key", key.get()),
+        Updates.set("value", value),
+        new UpdateOptions().upsert(true));
+    return null;
```

Comment:
The C* client requires us to be able to batch statements of different types (insert/delete) together and then execute them as a single batch on flush. We should change this API not to return `BoundStatement` which is specific to the Cassandra driver.

Reply:
Makes sense -- can we leave a TODO or do you plan to follow up on this immediately?

Reply:
I plan to do that relatively soon, it's bugging me quite a bit 😆 and I think it'll become more important when I do EOS support.